### PR TITLE
feat: autocomplete command

### PR DESCRIPTION
**Pull request by ForAeons**
### Changes
- Update Logic interface to include autocomplete
- Add autocomplete parsing to `AddressBookParser`
- Add `AutoComplete` Interface
- Add `AutoCompleteCommand` that autocompletes Commands
- Add key press listener in `CommandBox` that listens to TAB and autocompletes the given input string

### Remarks
- Review and merge #60 first
- Cannot pass ci because cannot test CommandBox ui (original ab3 app also did not test them)

### Screenshots
- No visible UI changes, just quality of life improvements

Resolves: #59
Resolves: #57

### CODE REVIEW

File: src/main/java/seedu/address/logic/autocomplete/AutoCompleteCommand.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+     * @param commands the commands to initialize the trie with
+     * @see seedu.address.logic.LogicManager
+     */
+    public static void initialize(String... commands) {
+        commandTrie = new Trie(commands);
+    }
+
+    @Override
+    public String getAutoComplete(String input) {
+        assert commandTrie != null;
```
Comment (belligerentbeagle) [inline src/main/java/seedu/address/logic/autocomplete/AutoCompleteCommand.java:23]: Great use of assert!

Comment (belligerentbeagle) [inline src/main/java/seedu/address/logic/autocomplete/AutoCompleteCommand.java:23]: Great use of assert!

File: src/test/java/seedu/address/logic/autocomplete/AutoCompleteCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+class AutoCompleteCommandTest {
+
+    @Test
+    void getAutoComplete() {
+        // Initialize the command trie with some random commands
+        AutoCompleteCommand.initialize("add", "delete", "clear", "edit", "exit", "find", "help", "list", "mark");
```
Comment (belligerentbeagle) [inline src/test/java/seedu/address/logic/autocomplete/AutoCompleteCommandTest.java:12]: I think add and delete should be addstu and delstu, right?

For completeness would you like to also add unmark? This one is up to you!

**Review by belligerentbeagle — CHANGES_REQUESTED** [review]
Thanks for doing up this feature! I think it's very useful to have this and interesting to see a trie being used IRL. Just a nit(?) in the segment for test, for consistency reasons!

File: src/test/java/seedu/address/logic/autocomplete/AutoCompleteCommandTest.java (hunk 2 above)
Comment (ForAeons) [reply src/test/java/seedu/address/logic/autocomplete/AutoCompleteCommandTest.java]: Thanks a lot for the review, I have made the changes

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 3]
```diff
@@ -56,6 +63,26 @@ private void setStyleToDefault() {
         commandTextField.getStyleClass().remove(ERROR_STYLE_CLASS);
     }
 
+    private void autoComplete(KeyEvent e) {
+        if (e.getCode() != KeyCode.TAB) {
+            return;
+        }
+
+        // Prevent the default Tab behavior
+        e.consume();
```
Comment (tituschewxj) [inline src/main/java/seedu/address/ui/CommandBox.java:72]: Maybe we can apply SLAP here and have another method `handleAutoComplete` that handles autocompletion, moving the lines after this into that method, and just have this method for checking the key code.
```suggestion
        if (e.getCode() == KeyCode.TAB) {
            handleAutoComplete();
            // Prevent the default Tab behavior
            e.consume();
        }
```

**Review by tituschewxj — CHANGES_REQUESTED** [review]
Thanks for implementing this feature! We may want to have more tests for this feature as right now some lines are only partially covered, but this is up to you, since we can still add tests in v1.4 but we can't fix any bugs that arise in v1.4.

File: src/main/java/seedu/address/logic/Logic.java [hunk 4]
```diff
@@ -25,6 +25,14 @@ public interface Logic {
      */
     CommandResult execute(String commandText) throws CommandException, ParseException;
 
+    /**
+     * Returns the autocomplete text based on the input.
+     * @param commandText The input as entered by the user.
```
[the remark this answers is not in the document]
Comment (ForAeons) [reply src/main/java/seedu/address/logic/Logic.java:31]: Resolved but didnt show file changed

File: src/main/java/seedu/address/logic/autocomplete/AutoComplete.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package seedu.address.logic.autocomplete;
+
+/**
+ * API of the Autocomplete component
+ */
+public interface AutoComplete {
+    /**
+     * Returns the autocomplete text based on the input.
+     * @param input The input as entered by the user.
```
[the remark this answers is not in the document]
Comment (ForAeons) [reply src/main/java/seedu/address/logic/autocomplete/AutoComplete.java:10]: Resolved but didnt show file changed

File: src/main/java/seedu/address/logic/autocomplete/AutoCompleteCommand.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Auto complete commands
+ */
+public class AutoCompleteCommand implements AutoComplete {
+    private static Trie commandTrie;
+
+    /**
+     * Initializes the command trie with the given commands. This method should be called once at
+     * the start of the initialization of LogicManager.
+     * @param commands the commands to initialize the trie with
```
[the remark this answers is not in the document]
Comment (ForAeons) [reply src/main/java/seedu/address/logic/autocomplete/AutoCompleteCommand.java:15]: Resolved but didnt show file changed

**Review by tituschewxj — APPROVED** [review]
LGTM! Thanks for the changes. Once all the comments have been resolved, then we can merge this.

**Review by belligerentbeagle — APPROVED** [review]
LGTM! Good work bro!

**Pull request merged**